### DISCUSSION TITLE
Implement documented ustruct.pack format codes

### DISCUSSION THREAD

**Discussion by byteit101** [micropython/micropython#12457] [Ideas]
Embedded systems often have to pull apart and put together packed binary structures, and `ustruct` is very helpful here, though there are 4 unimplemented format specifiers: `?` (bool), `c` (byte, fixed), `p` (byte-string, pascal), `x` (padding, in CircuitPython).

As the micropython ustruct documentation refers to the python struct document, I had assumed all the format specifiers would work, but this doesn't appear to be the case.

**Comment by jimmo** [comment]
> As the micropython ustruct documentation refers to the python struct document, I had assumed all the format specifiers would work, but this doesn't appear to be the case.

The micropython docs do explicitly list the supported data types.

Although `x` is missing... see https://github.com/micropython/micropython/pull/12059#issuecomment-1722081391

**Reply by byteit101** [reply]
Where? I don't see it on https://docs.micropython.org/en/v1.8/pyboard/library/ustruct.html

**Reply by jimmo** [reply]
That'd the docs for a very old version of micropython (v1.8, released in 2017)

Annoyingly the latest/nightly docs are currently offline, but the docs for the most recent release are here:

http://docs.micropython.org/en/v1.20.0/library/struct.html

**Reply by jimmo** [reply]
See https://github.com/orgs/micropython/discussions/12456#discussioncomment-7018647 about the latest docs being offline.

**Reply by byteit101** [reply]
Ah, still interesting that the first result for "micropython ustruct unpack" on google is that old

**Comment by jimmo** [comment]
> 4 unimplemented format specifiers: `?` (bool), `c` (byte, fixed), `p` (byte-string, pascal)

Agreed. It would probably be useful to add these.

> `x` (padding, in CircuitPython).

This was actually added a few days ago. (See the PR I linked above)

**Reply by byteit101** [reply]
I have already implemented ? on my branch, I can make a PR once I implement c and p

**Reply by byteit101** [reply]
I have finished and created PR https://github.com/micropython/micropython/pull/12458
